---
layout: page
title: About
---

We established MeDuSA, a mixed model-based method that leverages single-cell RNA-seq data as a reference to estimate cell-state abundance along a cell trajectory in bulk RNA-seq data. MeDuSA estimates cell abundance at each state while fitting the remaining cells of the same type individually as random effects, resulting in improved deconvolution accuracy over the compared methods. Extensive benchmark analyses demonstrate the effectiveness of MeDuSA, while applying it to cohort-level RNA-seq datasets reveals associations of cell-state abundances with disease or treatment conditions and cell-state-dependent genetic control of transcription. Our study provides a high-accuracy deconvolution tool to estimate cell-state abundance along a cell trajectory, and shows its utility in characterizing the dynamics of cell states in various biological processes.

Cite `MeDuSA`
-------------------
Liyang Song, Xiwei Sun, Ting Qi, Jian Yang

Contact
-------------------
If you have any questions about MeDuSA, please feel free to contact: songliyang@westlake.edu.cn.

Our group
-------------------
[Yang lab website](https://yanglab.westlake.edu.cn/)

Other tools developed by our group
-------------------
[A tool for Genome-wide Complex Trait Analysis (GCTA)](https://yanglab.westlake.edu.cn/software/gcta/#Overview)
[Summary-data-based Mendelian Randomization (SMR)](https://yanglab.westlake.edu.cn/software/smr/#Overview)
[Generalised Summary-data-based Mendelian Randomisaion (GSMR)](https://yanglab.westlake.edu.cn/software/gsmr/)
[OmicS-data-based Complex trait Analysis (OSCA)](https://yanglab.westlake.edu.cn/software/osca/#Overview)
